src/components/timeline/LayerLine.svelte

```diff
@@ -147,49 +151,86 @@
     }
   }
 
-  function resourcesToLinePoints(resources: Resource[], radius: number): LinePoint[] {
-    const points: LinePoint[] = [];
-    let id = 0;
+  function processResourcesToLinePoints(resources: Resource[]) {
+    if (typeof window === 'undefined') {
+      return;
+    }
+
+    window.cancelAnimationFrame(processingRequest);
+
+    points = [];
+    tempPoints = [];
+
+    processingRequest = window.requestAnimationFrame(() => resourcesToLinePoints(resources));
+  }
 
-    for (const resource of resources) {
+  function resourcesToLinePoints(
+    resources: Resource[],
+    resourceStartIndex = 0,
+    valueStartIndex = 0,
+    startId = 0,
+  ): void {
+    const startTime = performance.now();
+    let resourceIndex = resourceStartIndex;
+    let valueIndex = valueStartIndex;
+    let id = startId;
+
+    for (resourceIndex; resourceIndex < resources.length; ++resourceIndex) {
+      const resource = resources[resourceIndex];
       const { name, schema, values } = resource;
 
       if (schema.type === 'boolean') {
-        for (let i = 0; i < values.length; ++i) {
-          const value = values[i];
+        for (valueIndex; valueIndex < values.length; ++valueIndex) {
+          const value = values[valueIndex];
           const { x, y: yBoolean } = value;
           const y = yBoolean ? 1 : 0;
-          points.push({
+          tempPoints.push({
             id: id++,
             name,
-            radius,
             type: 'line',
             x,
             y,
           });
+
+          if (performance.now() - startTime > WORK_TIME_THRESHOLD) {
+            processingRequest = window.requestAnimationFrame(() =>
+              resourcesToLinePoints(resources, resourceIndex, valueIndex + 1, id),
+            );
+            return;
+          }
         }
+
+        valueIndex = 0;
       } else if (
         schema.type === 'int' ||
         schema.type === 'real' ||
         (schema.type === 'struct' && schema?.items?.rate?.type === 'real' && schema?.items?.initial?.type === 'real')
       ) {
-        for (let i = 0; i < values.length; ++i) {
-          const value = values[i];
+        for (valueIndex; valueIndex < values.length; ++valueIndex) {
+          const value = values[valueIndex];
           const { x } = value;
           const y = value.y as number;
-          points.push({
+          tempPoints.push({
             id: id++,
             name,
-            radius,
             type: 'line',
             x,
             y,
           });
+
+          if (performance.now() - startTime > WORK_TIME_THRESHOLD) {
+            processingRequest = window.requestAnimationFrame(() =>
+              resourcesToLinePoints(resources, resourceIndex, valueIndex + 1, id),
+            );
+            return;
+          }
         }
+
+        valueIndex = 0;
       }
     }
 
-    return points;
+    points = tempPoints;
```

Comment:
i'm probably just not understanding this, but what's the purpose of keeping both `points` and `tempPoints`?

Reply:
Basically at the start of processing I'm clearing out `points`, and then accumulating the results over multiple frames in `tempPoints` then assigning that to `points` when we're all done. So the `points` mutation causes one re-render after the processing is done. That being said, I was kind of wondering if maybe we do just mutate `points` each frame and we can like iteratively render the data as we process it ...

Reply:
Ooohh...I see now. My brain saw the `const points = []` declaration in the before snapshot and somehow understood that it was the change you introduced...